Object listing

hypeJunction opened the pull request:
**feature(views): improves usability of object listing views**

Improves usability of summary and full view listings by wrapping listing components in their own views, which should allow plugins to target listing elements via hooks and override individual parts without having to hijack an entire view.

Adds class selectors to make it easier to style object listings.

**feature(views): image block wrapper attributes can now be passed via `$vars`**

`page/components/image_block` now accepts additional attributes to be passed to the wrapper div with `$vars`.

hypeJunction committed:
feature(views): improves usability of object listing views

Improves usability of summary and full view listings by wrapping
listing components in their own views, which should allow plugins
to target listing elements via hooks and override individual parts
without having to hijack an entire view.

Adds class selectors to make it easier to style object listings.

hypeJunction committed:
feature(views): image block wrapper attributes can now be passed with $vars

page/components/image_block now accepts additional attributes to be passed to the
wrapper div with $vars.

hypeJunction commented on views/default/object/elements/full/body.php:
```diff
@@ -0,0 +1,13 @@
+	return;
+}
+?>
+<div class="elgg-listing-full-body"><?= $body ?></div>
```
This changes the markup, but I'd rather change it than have:
`html
<div class="elgg-content">
   <div class="elgg-content"></div>
   <div class="elgg-output"></div>
</div>
`
Makes it impossible to target the view.

mrclay replied:
Strawman: Introduce this view in 2.x, but make the markup change in 3.0?

hypeJunction replied:
I will be fine with that. I can just override the view.

hypeJunction commented on views/default/object/elements/summary/metadata.php:
```diff
@@ -0,0 +1,12 @@
+if (!$metadata) {
+	return;
+}
+echo $metadata;
```
Ideally we should wrap this as `<div class="elgg-list-summary-metadata">` and float it instead of floating the entity menu (makes it hard to display an entity menu elsewhere on the page without having to restyle it).

mrclay commented:
This is a good decomposition. Q is how much 3rd party themes will suffer and do we get much benefit without the markup changes (in 2.x).

hypeJunction commented on views/default/object/elements/summary.php:
```diff
@@ -1,59 +1,46 @@
-}
+$summary = $metadata . $title . $subtitle . $tags . $extensions . $content;
+
+echo elgg_format_element('div', [
```
This alters the markup as well.

hypeJunction commented:
Honestly don't know how a theme is managing to style anything with the current state of affairs.

hypeJunction commented:
I will update this to exclude markup changes in 2.x

hypeJunction committed:
chore(logs): logs an error if entity is missing from listing views

Log an error if entity is missing from objects/elements/full and object/elements/summary views

hypeJunction commented:
This is BC now. Will make other changes in master once this is merged. 
Added a third commit to log an error if an entity is missing.

mrclay commented on views/default/object/elements/summary.php:
```diff
@@ -1,59 +1,40 @@
+$title = elgg_extract('title', $vars, '');
+if ($title === '' && $entity instanceof ElggEntity) {
+	$vars['title'] = elgg_view('output/url', [
+		'text' => elgg_get_excerpt($entity->getDisplayName(), 100),
```
This 100 <- https://github.com/Elgg/Elgg/issues/9952

mrclay commented:
Besides the new views this just adds classes now, right?

hypeJunction commented:
I will update the theme sandbox and post before and after markup.

hypeJunction committed:
feature(developers): add custom attributes to image block sandbox view

Adds custom attributes to image block component in theme sandbox.

hypeJunction committed:
feature(developers): add object summary listing view to theme sandbox

hypeJunction commented:
Summary before:
```html
<ul class="elgg-menu elgg-menu-entity elgg-menu-hz elgg-menu-entity-default">
   <li class="elgg-menu-item-access"><span title="The access level" class="elgg-access elgg-access-private">Private</span></li>
   <li class="elgg-menu-item-edit"><a href="http://hackathon2.dev/theme-sandbox/edit/" title="Edit this" class="elgg-menu-content">Edit</a></li>
   <li class="elgg-menu-item-delete"><a href="http://hackathon2.dev/action/entity/delete?guid=&amp;__elgg_ts=1468136109&amp;__elgg_token=dv59yTG-B-CLh3FKJlOg6Q" title="Delete this" class="elgg-menu-content" data-confirm="Are you sure you want to delete this item?"><span class="elgg-icon-delete elgg-icon fa fa-times"></span></a></li>
</ul>
<h3><a href="http://hackathon2.dev/">Test Object</a></h3>
<div class="elgg-subtext">Listing subtitle</div>
<div class="clearfix">
   <ul class="elgg-tags">
      <li><span class="elgg-icon-tag elgg-icon fa fa-tag"></span></li>
      <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+1&amp;search_type=tags" rel="tag" value="tag 1">tag 1</a></li>
      <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+2&amp;search_type=tags" rel="tag" value="tag 2">tag 2</a></li>
      <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+3&amp;search_type=tags" rel="tag" value="tag 3">tag 3</a></li>
   </ul>
</div>
<div class="elgg-content">Sed scelerisque sagittis lorem. Phasellus sodales.
   Nulla urna justo, vehicula in, suscipit nec, molestie sed, tellus.
</div>
```

Summary after:
```html
<ul class="elgg-menu elgg-menu-entity elgg-menu-hz elgg-menu-entity-default">
   <li class="elgg-menu-item-access"><span title="The access level" class="elgg-access elgg-access-private">Private</span></li>
   <li class="elgg-menu-item-edit"><a href="http://hackathon2.dev/theme-sandbox/edit/" title="Edit this" class="elgg-menu-content">Edit</a></li>
   <li class="elgg-menu-item-delete"><a href="http://hackathon2.dev/action/entity/delete?guid=&amp;__elgg_ts=1468136496&amp;__elgg_token=V8DG4uA5GbxqoheydBdHlQ" title="Delete this" class="elgg-menu-content" data-confirm="Are you sure you want to delete this item?"><span class="elgg-icon-delete elgg-icon fa fa-times"></span></a></li>
</ul>
<h3 class="elgg-listing-summary-title"><a href="http://hackathon2.dev/">Test Object</a></h3>
<div class="elgg-listing-summary-subtitle elgg-subtext">Listing subtitle</div>
<div class="clearfix">
   <ul class="elgg-tags">
      <li><span class="elgg-icon-tag elgg-icon fa fa-tag"></span></li>
      <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+1&amp;search_type=tags" rel="tag" value="tag 1">tag 1</a></li>
      <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+2&amp;search_type=tags" rel="tag" value="tag 2">tag 2</a></li>
      <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+3&amp;search_type=tags" rel="tag" value="tag 3">tag 3</a></li>
   </ul>
</div>
<div class="elgg-listing-summary-content elgg-content">Sed scelerisque sagittis lorem. Phasellus sodales.
   Nulla urna justo, vehicula in, suscipit nec, molestie sed, tellus.
</div>
```

hypeJunction commented:
Full before:
```html
<div class="elgg-content clearfix theme-sandbox-full-listing">
   <div class="elgg-image-block clearfix">
      <div class="elgg-image"><a href="http://hackathon2.dev/"><img src="http://hackathon2.dev/cache/0/default/icons/default/small.png" alt="Test Object" width="40" height="40"></a></div>
      <div class="elgg-body">
         <ul class="elgg-menu elgg-menu-entity elgg-menu-hz elgg-menu-entity-default">
            <li class="elgg-menu-item-access"><span title="The access level" class="elgg-access elgg-access-private">Private</span></li>
            <li class="elgg-menu-item-edit"><a href="http://hackathon2.dev/theme-sandbox/edit/" title="Edit this" class="elgg-menu-content">Edit</a></li>
            <li class="elgg-menu-item-delete"><a href="http://hackathon2.dev/action/entity/delete?guid=&amp;__elgg_ts=1468136916&amp;__elgg_token=PUqarTmBJ6PS6fL6lyodvQ" title="Delete this" class="elgg-menu-content" data-confirm="Are you sure you want to delete this item?"><span class="elgg-icon-delete elgg-icon fa fa-times"></span></a></li>
         </ul>
         <h3><a href="http://hackathon2.dev/">Test Object</a></h3>
         <div class="elgg-subtext">Listing subtitle</div>
         <div class="clearfix">
            <ul class="elgg-tags">
               <li><span class="elgg-icon-tag elgg-icon fa fa-tag"></span></li>
               <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+1&amp;search_type=tags" rel="tag" value="tag 1">tag 1</a></li>
               <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+2&amp;search_type=tags" rel="tag" value="tag 2">tag 2</a></li>
               <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+3&amp;search_type=tags" rel="tag" value="tag 3">tag 3</a></li>
            </ul>
         </div>
      </div>
   </div>
   <div class="elgg-output">
      <p>Sed scelerisque sagittis lorem. Phasellus sodales.<br>Nulla urna justo, vehicula in, suscipit nec, molestie sed, tellus.</p>
   </div>
</div>
```

Full after:
```html
<div class="theme-sandbox-full-listing elgg-listing-full elgg-content clearfix">
   <div class="elgg-listing-full-header elgg-image-block clearfix">
      <div class="elgg-image"><a href="http://hackathon2.dev/"><img src="http://hackathon2.dev/cache/0/default/icons/default/small.png" alt="Test Object" width="40" height="40"></a></div>
      <div class="elgg-body">
         <ul class="elgg-menu elgg-menu-entity elgg-menu-hz elgg-menu-entity-default">
            <li class="elgg-menu-item-access"><span title="The access level" class="elgg-access elgg-access-private">Private</span></li>
            <li class="elgg-menu-item-edit"><a href="http://hackathon2.dev/theme-sandbox/edit/" title="Edit this" class="elgg-menu-content">Edit</a></li>
            <li class="elgg-menu-item-delete"><a href="http://hackathon2.dev/action/entity/delete?guid=&amp;__elgg_ts=1468137144&amp;__elgg_token=scgsQVp2TbiFA3et-A3WxA" title="Delete this" class="elgg-menu-content" data-confirm="Are you sure you want to delete this item?"><span class="elgg-icon-delete elgg-icon fa fa-times"></span></a></li>
         </ul>
         <h3 class="elgg-listing-summary-title"><a href="http://hackathon2.dev/">Test Object</a></h3>
         <div class="elgg-listing-summary-subtitle elgg-subtext">Listing subtitle</div>
         <div class="clearfix">
            <ul class="elgg-tags">
               <li><span class="elgg-icon-tag elgg-icon fa fa-tag"></span></li>
               <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+1&amp;search_type=tags" rel="tag" value="tag 1">tag 1</a></li>
               <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+2&amp;search_type=tags" rel="tag" value="tag 2">tag 2</a></li>
               <li class="elgg-tag"><a href="http://hackathon2.dev/search?q=tag+3&amp;search_type=tags" rel="tag" value="tag 3">tag 3</a></li>
            </ul>
         </div>
      </div>
   </div>
   <div class="elgg-output">
      <p>Sed scelerisque sagittis lorem. Phasellus sodales.<br>Nulla urna justo, vehicula in, suscipit nec, molestie sed, tellus.</p>
   </div>
</div>
```

hypeJunction committed:
feature(developers): add object full listing to theme sandbox

hypeJunction committed:
feature(views): object summary listing now accepts an icon

Allow passing an icon to the object summary listing view in order to wrap content
consistently with .elgg-list-summary image block

hypeJunction commented on views/default/object/elements/full/body.php:
```diff
@@ -0,0 +1,13 @@
+	return;
+}
+
+echo $body;
```
#9978

mrclay commented:
LGTM, but let's get another

hypeJunction commented on views/default/page/components/image_block.php:
```diff
@@ -15,38 +15,41 @@
 
 if ($alt_image) {
-	$alt_image = "<div class=\"elgg-image-alt\">$alt_image</div>";
+	$alt_image = elgg_format_element([
```
This causes fatal error

hypeJunction replied:
Fixed. Also added image block with image alt to the sandbox

hypeJunction committed:
chore(views): passing alt image to image block no longer causes errors

Also adds image block with alt image to the theme sandbox

hypeJunction commented:
@jdalsem Could you take a look?

jdalsem commented:
LGTM... good to have more control over summary parts...
